File: 6-2-1.java
import java.util.*;

public class Main {
    static class Pair {
        int first;
        int second;
        
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
    
    static class Node implements Comparable<Node> {
        int index;
        long distance;
        
        Node(int index, long distance) {
            this.index = index;
            this.distance = distance;
        }
        
        public int compareTo(Node other) {
            return Long.compare(this.distance, other.distance);
        }
    }
    
    static final long INF = (long)1e18;
    static List<Pair>[] E;
    
    static void dijkstra(int start, int end, List<Long> dist) {
        dist.set(start, 0L);
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(start, 0L));
        
        while (!pq.isEmpty()) {
            Node node = pq.poll();
            long cost = node.distance;
            int here = node.index;
            
            if (dist.get(here) < cost) continue;
            
            for (Pair edge : E[here]) {
                int there = edge.first;
                long nextDist = cost + edge.second;
                
                if (dist.get(there) > nextDist) {
                    dist.set(there, nextDist);
                    pq.add(new Node(there, nextDist));
                }
            }
        }
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int M = sc.nextInt();
        int X, Z;
        int u, v, w;
        
        E = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            E[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < M; i++) {
            u = sc.nextInt();
            v = sc.nextInt();
            w = sc.nextInt();
            E[u].add(new Pair(v, w));
        }
        
        X = sc.nextInt();
        Z = sc.nextInt();
        
        List<Long> dist = new ArrayList<>(N + 1);
        for (int i = 0; i <= N; i++) {
            dist.add(INF);
        }
        
        dijkstra(X, Z, dist);
        
        if (dist.get(Z) >= INF) {
            dist.set(Z, -1L);
        }
        
        System.out.println(dist.get(Z));
    }
}

File: 6-2-1.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<int, int> pii;
const ll INF = ll(1e18);
vector<pii> E[100004];
void dijkstra(int start, int end, vector<ll>& dist) {
	dist[start] = 0;
	priority_queue<pair<ll, int>> pq;
	pq.push({ 0, start });
	while (!pq.empty()) {
		ll cost = -pq.top().first;
		int here = pq.top().second;
		pq.pop();

		if (dist[here] < cost) continue;

		for (int i = 0; i < (int)E[here].size(); ++i) {
			int there = E[here][i].first;
			ll next_dist = cost + E[here][i].second;

			if (dist[there] > next_dist) {
				dist[there] = next_dist;
				pq.push(make_pair(-next_dist, there));
			}
		}
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N, M, X, Z;
	int u, v, w;
	cin >> N >> M;
	while (M-- > 0) {
		cin >> u >> v >> w;
		E[u].push_back({ v,w });
	}
	cin >> X >> Z;
	vector<ll> dist(N + 1, INF);
	dijkstra(X, Z, dist);
	if (dist[Z] >= INF) dist[Z] = -1;
	cout << dist[Z];
}